File: flutter_app/test/task_repository_test.dart
import 'package:drift/native.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:lifetrace_execute/data/local/app_database.dart';
import 'package:lifetrace_execute/data/repository/entity_link_repository.dart';
import 'package:lifetrace_execute/data/repository/task_repository.dart';
import 'package:lifetrace_execute/domain/task/execution_task.dart';

void main() {
  late AppDatabase database;
  late DriftTaskRepository repository;

  setUp(() {
    database = AppDatabase(NativeDatabase.memory());
    repository = DriftTaskRepository(database);
  });

  tearDown(() async {
    await database.close();
  });

  test('create/update/delete keep task and outbox in local-first parity', () async {
    final created = await repository.createTask(
      userId: 'user-1',
      deviceId: 'device-1',
      title: '  Flutter Task  ',
      description: 'local first',
      priority: ExecutionTaskPriority.high,
    );

    expect(created.title, 'Flutter Task');
    expect(await database.select(database.tasks).get(), hasLength(1));
    var outbox = await database.select(database.syncOutbox).get();
    expect(outbox, hasLength(1));
    expect(outbox.single.operation, 'upsert');
    expect(outbox.single.entityType, DriftTaskRepository.entityType);

    final completed = await repository.updateTask(
      task: created,
      deviceId: 'device-1',
      status: ExecutionTaskStatus.done,
    );

    expect(completed.localVersion, 2);
    expect(completed.completedAt, isNotNull);
    expect((await database.select(database.tasks).get()).single.status, 'done');
    outbox = await database.select(database.syncOutbox).get();
    expect(outbox, hasLength(2));

    await repository.deleteTask(userId: 'user-1', taskId: created.id);

    expect(await database.select(database.tasks).get(), isEmpty);
    outbox = await database.select(database.syncOutbox).get();
    expect(outbox, hasLength(3));
    expect(outbox.last.operation, 'delete');
  });

  test('deleting a task tombstones task links without deleting the child task',
      () async {
    final parent = await repository.createTask(
      userId: 'user-1',
      deviceId: 'device-1',
      title: 'Parent',
    );
    final child = await repository.createTask(
      userId: 'user-1',
      deviceId: 'device-1',
      title: 'Child',
    );
    final links = DriftEntityLinkRepository(database);
    await links.createLink(
      userId: 'user-1',
      deviceId: 'device-1',
      sourceType: DriftTaskRepository.entityType,
      sourceId: parent.id,
      targetType: DriftTaskRepository.entityType,
      targetId: child.id,
      relationType: 'subtask',
    );

    expect(await database.select(database.entityLinks).get(), hasLength(1));

    await database.delete(database.syncOutbox).go();
    await repository.deleteTask(userId: 'user-1', taskId: parent.id);

    final tasks = await database.select(database.tasks).get();
    expect(tasks.map((item) => item.id), contains(child.id));
    expect(tasks.map((item) => item.id), isNot(contains(parent.id)));
    expect(await database.select(database.entityLinks).get(), isEmpty);

    final outbox = await database.select(database.syncOutbox).get();
    expect(outbox, hasLength(2));
    expect(
      outbox.map((item) => item.entityType),
      containsAll(<String>[
        DriftEntityLinkRepository.entityType,
        DriftTaskRepository.entityType,
      ]),
    );
    expect(outbox.every((item) => item.operation == 'delete'), isTrue);
  });

}

File: app/src/test/java/com/lifetrace/execute/data/repository/TaskRepositoryTest.kt
package com.lifetrace.execute.data.repository

import android.app.Application
import android.content.Context
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.lifetrace.execute.data.local.LifeTraceExecuteDatabase
import com.lifetrace.execute.domain.task.ExecutionTaskPriority
import com.lifetrace.execute.domain.task.ExecutionTaskStatus
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [35], application = Application::class)
class TaskRepositoryTest {
    private lateinit var database: LifeTraceExecuteDatabase
    private lateinit var repository: TaskRepository

    @Before
    fun setUp() {
        val context = ApplicationProvider.getApplicationContext<Context>()
        database = Room.inMemoryDatabaseBuilder(
            context,
            LifeTraceExecuteDatabase::class.java,
        ).allowMainThreadQueries().build()
        repository = TaskRepository(database)
    }

    @After
    fun tearDown() {
        database.close()
    }

    @Test
    fun createPersistsEntityAndOutboxTogether() = runBlocking {
        val task = repository.createTask(
            userId = "user-1",
            deviceId = "device-a",
            title = "  Local first task  ",
            description = "  persists before network  ",
            projectId = "project-1",
            priority = ExecutionTaskPriority.HIGH,
            dueAt = "2026-09-02T12:00:00Z",
        )

        val stored = database.dao().getTask("user-1", task.id)
        val outbox = database.dao().firstOutboxForEntity(
            "user-1",
            TaskRepository.ENTITY_TYPE,
            task.id,
        )

        assertNotNull(stored)
        assertEquals("Local first task", stored!!.title)
        assertEquals("persists before network", stored.description)
        assertEquals("project-1", stored.projectId)
        assertNotNull(outbox)
        assertEquals("upsert", outbox!!.operation)
        assertEquals("0", outbox.baseServerVersion)
        assertTrue(outbox.dependenciesJson.contains("project-1"))
        assertEquals(1, database.dao().observePendingOutboxCount("user-1").first())
    }

    @Test
    fun updateIncrementsLocalVersionAndQueuesAnotherChange() = runBlocking {
        val original = repository.createTask(
            userId = "user-1",
            deviceId = "device-a",
            title = "Original",
        )

        val updated = repository.updateTask(
            task = original,
            deviceId = "device-b",
            title = "Updated",
            status = ExecutionTaskStatus.DONE,
            priority = ExecutionTaskPriority.URGENT,
        )

        val stored = database.dao().getTask("user-1", original.id)
        assertEquals(2, updated.localVersion)
        assertEquals(2, stored!!.localVersion)
        assertEquals("done", stored.status)
        assertEquals("urgent", stored.priority)
        assertNotNull(stored.completedAt)
        assertEquals("device-b", stored.modifiedByDevice)
        assertEquals(2, database.dao().observePendingOutboxCount("user-1").first())
    }

    @Test
    fun deleteRemovesVisibleEntityButKeepsDeleteIntentInOutbox() = runBlocking {
        val task = repository.createTask(
            userId = "user-1",
            deviceId = "device-a",
            title = "Delete me",
        )

        repository.deleteTask("user-1", task.id)

        assertNull(database.dao().getTask("user-1", task.id))
        assertEquals(2, database.dao().observePendingOutboxCount("user-1").first())
    }

    @Test
    fun blankTitleIsRejectedBeforeAnyDatabaseWrite() {
        assertThrows(IllegalArgumentException::class.java) {
            runBlocking {
                repository.createTask(
                    userId = "user-1",
                    deviceId = "device-a",
                    title = "   ",
                )
            }
        }
    }
}
